pkg/resource/deploy/step_generator.go

```diff
@@ -167,7 +223,28 @@ func (sg *stepGenerator) GenerateSteps(event RegisterResourceEvent) ([]Step, err
 		}, nil
 	}
 
-	// Case 2: hasOld
+	// Case 2: wasExternal
```

Comment:
Hmm, this is getting subtle.

I get that relinquish and taking ownership are super interesting and powerful conceptual operations, I just get nervous that automatically performing them when you change your program from `new` to `get` (or vice versa) is very subtle. I would have almost preferred explicit options to perform these actions and to by default do what the natural composition of operations would be.

Also, unless I'm mistaken, it really isn't taking ownership of an existing resource, because we're creating a replacement?

Reply:
> I get that relinquish and taking ownership are super interesting and powerful conceptual operations, I just get nervous that automatically performing them when you change your program from new to get (or vice versa) is very subtle. I would have almost preferred explicit options to perform these actions and to by default do what the natural composition of operations would be.

I understand the nervousness, but I'll admit that as soon as we introduce this notion of first class reads, we kind of have to provide semantics for these cases, and I think the semantics here are the most expected ones if you think about what a user might expect to happen when making these changes.

Having these behave in a surprising way and offering another tool to do what is expected doesn't necessarily feel better, even if it could possibly simplify the model here.

I think the complexity here is somewhat fundamental to having a 1st class notion of "external" resources in a resource model.

Reply:
Erroring in this case unless you pass `convertExistingResource: true` seems reasonable.  Though not clear that this helps with the issue of deleting the resource if you got the ID wrong.  And overall, feels like it puts a speed bump on the wrong side of the usability here - the bad thing is just as easy, the good thing is just harder.

I'm sympathetic to composition arguments, but once we introduce new first class notion of external resources, the state transition matrix just isn't necessarily orthogonally compositional.  

We should probably have offline discussion on this to make sure we are all on the same page - I know @swgillespie and @pgavlin spent several hours working through this exact topic - and then another hour with me earlier this week.

Reply:
One of the other issues with an approach like `convertExistingResource: true` is that once that program has executed, it may immediately be incorrect, assuming that `convertExistingResource` expects the current state to be non-external. This would then require the user to make another code change in order to make forward progress.

Reply:
Early on when doing this PR I thought that we might want to consider some sort of CLI gesture when doing one of these more advanced operations. I agree with Pat that this becomes troublesome when the gesture enters the code because we're drifting away from desired state configuration semantics. Luke introduced a test that's been very helpful when we were talking last week - any two sequences of steps through our system composed in any order should get us to the same desired state ("path independence").